API 2.0 con nuevos requisitos

miguescri opened the pull request:
Añadidos los requisitos sobre autores y álbunes, entre otros. Resultado de la corrección de requisitos hecha en #2

miguescri committed:
API 2.0 con nuevos requisitos (#15)

Añadidos los requisitos sobre autores y álbunes, entre otros.

miguescri commented:
COMENTARIO DE @robertoccu EN EL PULL REQUEST PREVIO:

Parece estar todo bien, solo una cosa. Si al obtener un usuario se obtiene toda la información de cada uno de sus amigos, ¿no se creará un bucle al recibir también la información de los amigos de los amigos hasta el infitino? Igual con devolver ID y nombre de cada amigo sería suficiente.

EDIT: Como indique en el issue #2 faltarían fotos de artistas/grupos. Author.

miguescri commented:
Tienes razón con lo del bucle infinito de amigos; de hecho la interfaz que muestra [SwaggerHub](https://app.swaggerhub.com/apis/miguescri/ProySoft/2.0.0#/) ya se enfrenta a ese problema poniendo el campo a nulo.

¿Te parece bien si lo dejo tal y como está pero en la definición del método indico que el campo "friends" de los amigos del usuario estará vacío?

Ejemplo de respuesta recibida al pedir el perfil de Stallman:

{
  "id": 11111,
  "username": "rms",
  "name": "Richard Stallman",
  "bio": "¡Larga vida al Open Source!",
  "mail": "rms@gnu.org",
  "friends": [
    {
      "id": 12345,
      "username": "juan",
      "name": "juan",
      "bio": "",
      "friends": [
        null
      ]
    },
    {
      "id": 9876,
      "username": "damian",
      "name": "damián",
      "bio": "",
      "friends": [
        null
      ]
    },
  ]
}

miguescri committed:
Correcciones API 2.0 (#17)

* Añadido endpoint de imagen de autor

* Corregidas descripciones de métodos

Se han corregido erratas en las descripciones de algunos métodos y se
ha especificado el comportamiento del campo "friends": al recibir un
perfil de usuario se obtienen los perfiles de sus amigos, pero estos no
arrastran a su vez los perfiles de sus amigos, sino que tienen valor NULL
en el campo "friends".

miguescri commented:
Añadidas correcciones al pull request a través de #17

robertoccu commented:
@miguescri Serviría, ahi ya lo dejaría a tu saber según como os sea más cómodo en back-end. En modelo, en principio, de los amigos cogeremos solo nombre para mostrarlo e ID para conseguir el resto de información (incluido amigos) al entrar al prefil.

miguescri commented:
Hombre, haz uso de lo que te damos. Si he metido info de los amigos al rescatar un perfil es para que os ahorréis posteriores llamadas innecesarias a esos perfiles de amigo. Con esta definición de API sólo necesitaríais rellamadas en caso de necesitar la lista de amigos del amigo.

miguescri commented:
Acabo de darme cuenta de que, aunque lo había pensado, no he añadido ni a ProfileItem ni a AccountItem las listas de reproducción del usuario. Eso ahorraría aún más llamadas y reduciría la latencia de la aplicación.

Lo único, habría que hacer lo mismo que en el caso anterior, y poner a NULL el campo de listas de reproducción de los amigos.

miguescri committed:
Añadidas listas de reproducción a perfiles de propietarios

Ahora tanto los ProfileItem como los Account Item contiene un array con
las listas de reproducción de las que es propietario el usuario.

miguescri commented:
Añadidas.

miguescri committed:
Corrección README

robertoccu commented:
@miguescri Sobre lo de usar la información, me refería a que cuando tu recibes el perfil de una persona, solo ves el nombre de sus amigos, el resto de información no se muestra hasta que entras en el perfil de ese amigo. Aunque quisieramos, no tendría mucho sentido mostrar más información de los amigos (quizá la biografía).

Al entrar en el perfil del amigo habrá que llamar a la API de nuevo para obtener los amigos de ese amigo.

---

Por otra parte, acabo de fijarme que dentro de la API, el obtener el archivo de una canción está dentro de Public. ¿No debería estar dentro de User? Por eso de que el usuario tenga que estar registrado para usar el servicio.

miguescri commented:
En la categoría Users he metido los métodos que sólo puede hacer un usuario en concreto (modificar una lista de tu propiedad) o que devuelven un resultado diferente para cada usuario (obtener datos de Account). Todo lo que queda fuera de esas categorías (y de Admin, claro) lo he englobado en Public. En el caso de modificar el get de canciones para requerir login sólo haría falta añadir una posible respuesta de error 401 siguiendo en la misma categoría.

Respecto a lo de estar registrado para escuchar una canción, no lo había visto en los requisitos. Siguiendo esa línea de pensamiento, ¿también haría falta estar logeado para ver las playlist y los perfiles de los usuarios, autores, álbunes y demás?

robertoccu commented:
Era un requisito que se necesitaba una cuenta para usar el servicio, pero si que es verdad que nunca se especificó en la lista de requisitos ni a nadie se le ocurrió en su momento para incorporarlo, como otras tantas cosas que están surgiendo. "Usar" sería escuchar canciones, el resto de cosas, como ver perfiles, playlists, etc... lo dejaría público como hacen servicios como Spotify o Google Play Music.

miguescri committed:
Corregida errata en una etiqueta

miguescri commented:
He hablado con @robertoccu y, puesto que sólo tenemos canciones sin copypright, nos ahorramos la necesidad de autentificación para obtener los mp3.

robertoccu reviewed (approved):
Todo bien

miguescri commented:
El tema de las dependencias circulares de las listas de amigos no quedó solucionado con mi enfoque de dejar por escrito la restricción, ya que el código generado por Swagger (al menos el de Python), por lo que he modificado la API para solucionarlo en el pull #18
